internal/service/admin/adminservice/signin.go

```diff
@@ -30,6 +30,28 @@ func (s *adminService) Signin(ctx context.Context, u *ent.Admin) error {
 				return apistatus.NewAuthorization("Invalid email and password combination")
 			}
 
+			/*
+				Angus: Error handling is such a huge topic! This isn't necessarily a bad way of handling errors, but be
```

Comment:
I'm glad you think so! There are many approaches, so I encourage you to experiment. For example, your domain errors could satisfy some interface such as:

```go
type DisplayError interface {
    Display(userFacingFieldName string) string
}
```

Then in your handlers, you could simply do a type assertion to check whether an error is a `DisplayError`, and call its `Display` method to add its user-friendly error message to your JSON response object.

There's a trade-off here, though. Because the domain and the transport layer may refer to the same fields by different names (for example, the public API may change over time, but the domain may not), a `DisplayError` doesn't know what to call the malformed field in its message. The domain field name might not make sense to a user, who submitted the incorrect value under a totally different key! That's why it has to be passed in from the handler as `userFacingFieldName` when the error message is returned to the user.

However, what if this particular `DisplayError` doesn't refer to a specific field? Or any field? Or what if it refers to multiple fields? These cases are really hard to handle on the domain side, because they require too much transport layer knowledge.

That's why I prefer mapping many different types of domain errors to error messages in my handlers. It takes more boilerplate, and sometimes duplicates domain error messages that _would_ make sense to the user, but you guarantee that your messages will ALWAYS make sense to the user, and you keep all transport concerns out of the domain.

People don't put enough thought into error handling when they start a new service. If you get the pattern correct at the beginning, you'll end up with fewer bugs in production, it'll be much easier to debug, and the mapping from domain error to user-friendly message in the transport layer provides a really clear list of every error you intend to handle. Then you know that anything not in that list is a 500!